pkg/ingester/ingester.go

```diff
@@ -241,13 +250,48 @@ func (i *Ingester) getOrCreateInstance(instanceID string) *instance {
 
 // Query the ingests for log streams matching a set of matchers.
 func (i *Ingester) Query(req *logproto.QueryRequest, queryServer logproto.Querier_QueryServer) error {
-	instanceID, err := user.ExtractOrgID(queryServer.Context())
+	// initialize stats collection for ingester queries and set grpc trailer with stats.
+	ctx := stats.NewContext(queryServer.Context())
+	defer stats.SendAsTrailer(ctx, queryServer)
+
+	instanceID, err := user.ExtractOrgID(ctx)
 	if err != nil {
 		return err
 	}
 
 	instance := i.getOrCreateInstance(instanceID)
-	return instance.Query(req, queryServer)
+	itrs, err := instance.Query(ctx, req)
+	if err != nil {
+		return err
+	}
+
+	if i.cfg.QueryStore {
+		start := req.Start
+		end := req.End
+		if i.cfg.QueryStoreMaxLookBackPeriod != 0 {
+			oldestStartTime := time.Now().Add(-i.cfg.QueryStoreMaxLookBackPeriod)
+			if oldestStartTime.After(req.Start) {
+				start = oldestStartTime
+			}
+		}
+
+		if start.Before(end) {
+			storeRequest := recreateRequestWithTime(req, start, end)
```

Comment:
I was not clear, but my idea was there's a lot of if here and I don't think you have a test covering each of those.

I think you should create a function like this:

```golang
// buildStoreRequest returns a store request from a ingester request, return nit if no request should be made based on configuration. The request may be truncated due to QueryStoreMaxLookBackPeriod, explain QueryStoreMaxLookBackPeriod purpose here
func buildStoreRequest(config Config,req *logproto.QueryRequest) (logql.SelectParams,error)
```

This has the advantages of making it easier to test. Now you can write tests with a combination of requests and configs and set some expectation has to when a request will be fired and how/which.

Now usage wise it would look like this:

```golang
if req, err:= buildStoreRequest(req,i.cfg);err !=nil && req != nil {
  storeItr, err := i.store.LazyQuery(ctx,req)
  if err != nil {
    return err
  }
 itrs = append(itrs, storeItr)
}
```